src/controllers/utils/advanced-search-validation.ts

```diff
@@ -0,0 +1,94 @@
+                    if (!validDate) {
+                        throw Error(INVALID_DATE_ERROR_MESSAGE);
+                    }
+                    const toDate = req.query?.incorporatedTo;
+                    if (checkStringNotNullOrEmpty(toDate)) {
+                        if (checkDateBeforeInitial(toDate, date)) {
+                            throw Error(TO_DATE_BEFORE_FROM_DATE);
+                        }
+                    }
+                    if (checkIfDateInFuture(date)) {
+                        throw Error(INCORPORATION_DATE_IN_FUTURE);
+                    }
+                }
+                return true;
+            }),
+        check(INCORPORATED_TO_FIELD)
+            .custom((date, { req }) => {
+                if (checkStringNotNullOrEmpty(date)) {
+                    const validDate = validateDate(date);
+                    if (!validDate) {
+                        throw Error(INVALID_DATE_ERROR_MESSAGE);
+                    }
+                    const fromDate = req.query?.incorporatedFrom;
+                    if (checkStringNotNullOrEmpty(fromDate)) {
+                        if (checkDateBeforeInitial(date, fromDate)) {
+                            throw Error(TO_DATE_BEFORE_FROM_DATE);
+                        }
+                    }
+                    if (checkIfDateInFuture(date)) {
+                        throw Error(INCORPORATION_DATE_IN_FUTURE);
+                    }
+                }
+                return true;
+            })
+    ];
+
+export const validate = (validationErrors) => {
+    let incorporatedFromError;
+    let incorporatedToError;
+    const validationErrorList = validationErrors.array({ onlyFirstError: true }).map((error) => {
+        const govUkErrorData: GovUkErrorData = createGovUkErrorData(error.msg, "#" + error.param, true, "");
+        switch (error.param) {
+        case INCORPORATED_FROM_FIELD:
+            incorporatedFromError = govUkErrorData;
+            break;
+        case INCORPORATED_TO_FIELD:
+            incorporatedToError = govUkErrorData;
+            break;
+        }
+        return govUkErrorData;
+    });
+    return {
+        errorList: validationErrorList,
+        incorporatedFromError,
+        incorporatedToError
+    };
+};
+
+function checkDateBeforeInitial (initialDate: string, checkDate: string) {
+    const firstMoment = moment(initialDate, "DD/MM/YYYY");
+    const secondMoment = moment(checkDate, "DD/MM/YYYY");
+    const firstDate = firstMoment.toDate();
+    const secondDate = secondMoment.toDate();
+    return firstDate < secondDate;
+}
+
+function checkStringNotNullOrEmpty (checkString: string) {
+    return checkString !== null && checkString !== undefined && checkString.length > 0;
+}
+
+function checkIfDateInFuture (date: string) {
+    const dateMoment = moment(date, "DD/MM/YYYY");
+    const checkDate = dateMoment.toDate();
+    const now = new Date();
+    return now < checkDate;
+}
```

Comment:
I think adding the return type `: boolean` to these will make it easier at a quick glance to see what these functions do. Also another common standard for booleans is to use `is` as a prefix, for example:
`isDateBeforeInitial`
`isStringNotNullOrEmpty`
`isDateInFuture`

Reply:
Changed as per Cai's comments